### PR TITLE
Gtk4

### PR DESCRIPTION
**Pull request by leolost2605**
Todo:

- [x] investigate DND with Popover (PlaylistPopover) causing some really weird issues (e.g. the popover can't be closed and will stay on top even if other applications get focused)
- [x] The posters have margins for some reason
- [x] Size of PreviewPopover
- [x] Open File throws criticals

Pretty much ready however there is the issue that dragging and dropping in the `PlaylistPopover` causes some really weird issues (e.g. the popover can't be closed and will stay on top even if other applications get focused, also the main window can't be focused and doesn't respond to input). I'm not 100% sure but I think this might be an issue in mutter/gala/gtk so I don't think we can do anything about it currently (this issue also exists in the gtk4 widget factory; also there seem to be problems with DND in popovers on wayland (see [this issue](https://gitlab.gnome.org/GNOME/mutter/-/issues/1681))). Maybe worth noting  is that when `autohide=false` this doesn't occur however IMO that's not a perfect solution. Maybe we can use a separate window instead of a popover or disable DND in the playlist entirely for now?

Also this uses [this gst plugin](https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/tree/main/video/gtk4) written in rust. It's currently by far the easiest way to get GTK4 integration with gstreamer. It needs the rust SDK though which is why CI is failing, any tips here?

### CODE REVIEW

**Comment by leolost2605** [comment]
Hmm I just need a way to install the rust SDK in CI because it isn't in appcenter. Maybe we can add flathub as an additional source?

**Comment by davidmhewitt** [comment]
Yeah, you can just add this to the GitHub action:

```
      - name: Install Rust SDK extension
        run: |
          flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo
          flatpak install -y --arch=${{matrix.arch}} org.freedesktop.Sdk.Extension.rust-stable//22.08
```

It's only needed for the build, so user's don't need it installed to run.

File: src/Widgets/Library/EpisodesPage.vala [hunk 1]
```diff
@@ -240,9 +237,9 @@ public class Audience.EpisodesPage : Gtk.Box {
             }
         }
 
-        var deck = (Hdy.Deck) get_ancestor (typeof (Hdy.Deck));
-        if (deck.visible_child == this && items.get_n_items () == 0) {
-            deck.navigate (Hdy.NavigationDirection.BACK);
+        var Leaflet = (Adw.Leaflet) get_ancestor (typeof (Adw.Leaflet));
+        if (Leaflet.visible_child == this && items.get_n_items () == 0) {
+            Leaflet.navigate (Adw.NavigationDirection.BACK);
```
Comment (danirabbit) [inline src/Widgets/Library/EpisodesPage.vala:242]: variable names should be lowercase :)

Comment (leolost2605) [reply src/Widgets/Library/EpisodesPage.vala]: Ah yeah good old replace all :) 
Now fixed.

**Comment by danirabbit** [comment]
@leolost2605 Yeah let's just pull out the Drag n Drop stuff for now since it's broken. I'd really like to make some larger design changes in this app anyways in future branches :)

File: src/Widgets/Library/LibraryItem.vala [hunk 2]
```diff
@@ -74,25 +74,39 @@ public class Audience.LibraryItem : Gtk.FlowBoxChild {
             margin_end = 12
         };
 
-        context_menu = new Gtk.Menu ();
+        var move_to_trash = new Gtk.Button.with_label (_("Move to Trash"));
```
Comment (danirabbit) [inline src/Widgets/Library/LibraryItem.vala:77]: Need to left align these labels

Comment (leolost2605) [reply src/Widgets/Library/LibraryItem.vala]: Done.
Is there a "good" way to position the popover at the bottom right? Or should it be kept in the middle? Currently it's done via setting the offset which sometimes causes some flickering

Comment (danirabbit) [reply src/Widgets/Library/LibraryItem.vala]: Done! You can use halign for this :)

**Comment by danirabbit** [comment]
@leolost2605 fixed that menu align. If you think this is ready, I think let's get it merged since we just did a release and there's plenty of time to fix up any rough corners we find while testing :)

**Comment by leolost2605** [comment]
@danirabbit sure as far as I'm concerned this is ready. Also you said something about some design changes, if you've got some mockups I could work on implementing them :) Since music library is unfortunately forced on hold until OS8...

**Review by danirabbit — APPROVED** [review]
🚀 let's goooo

**Pull request merged**